modules/extension/app-schema/app-schema/src/main/java/org/geotools/data/complex/filter/FeatureChainedAttributeVisitor.java

```diff
@@ -0,0 +1,573 @@
+ * feature types / attributes must be traversed to reach the attribute identified by the provided {@link PropertyName} expression.
+ * 
+ * <p>
+ * The provided {@link FeatureTypeMapping} object is regarded as the root mapping against which the expression is evaluated.
+ * </p>
+ * 
+ * <p>
+ * The nested attribute mappings are returned as a list of {@link FeatureChainLink} objects; the first one in the list always refers to the root
+ * mapping.
+ * </p>
+ * 
+ * @author Stefano Costa, GeoSolutions
+ *
+ */
+public class FeatureChainedAttributeVisitor extends DefaultExpressionVisitor {
+
+    private static final Logger LOGGER = Logging.getLogger(FeatureChainedAttributeVisitor.class);
+
+    private FeatureTypeMapping rootMapping;
+
+    private List<FeatureChainedAttributeDescriptor> attributes;
+
+    public FeatureChainedAttributeVisitor(FeatureTypeMapping root) {
+        if (root == null) {
+            throw new NullPointerException("root mapping is null");
+        }
+        this.attributes = new ArrayList<>();
+        this.rootMapping = root;
+    }
+
+    @Override
+    public Object visit(PropertyName expression, Object data) {
+        if (expression == null) {
+            throw new NullPointerException("expression is null");
+        }
+        Feature feature = null;
+        if (data != null && !(data instanceof Feature)) {
+            feature = (Feature) data;
+        }
+
+        // reset outcome of the visit
+        attributes = new ArrayList<>();
+
+        try {
+            walkXPath(expression.getPropertyName(), feature);
+        } catch (IOException e) {
+            throw new RuntimeException(
+                    "Exception occurred splitting XPath expression into mapping steps", e);
+        }
+
+        return getFeatureChainedAttributes();
+    }
+
+    void walkXPath(String xpath, Feature feature) throws IOException {
+        FeatureTypeMapping currentType = rootMapping;
+        StepList currentXPath = XPath.steps(rootMapping.getTargetFeature(), xpath,
+                rootMapping.getNamespaces());
+        FeatureChainedAttributeDescriptor attrDescr = new FeatureChainedAttributeDescriptor();
+        walkXPathRecursive(currentXPath, currentType, attrDescr, feature);
+    }
+
+    private void walkXPathRecursive(StepList currentXPath, FeatureTypeMapping currentType,
+            FeatureChainedAttributeDescriptor attrDescr, Feature feature) throws IOException {
+        List<NestedAttributeMapping> currentAttributes = currentType.getNestedMappings();
+        boolean searchIsOver = true;
+        for (NestedAttributeMapping nestedAttr : currentAttributes) {
+            StepList targetXPath = nestedAttr.getTargetXPath();
+
+            if (currentXPath.startsWith(targetXPath)) {
+                if (nestedAttr.isConditional() && feature == null) {
+                    LOGGER.fine("Conditional nested mapping found, but no feature to evaluate "
```

Comment:
Some more information here including the xpath? This would help users debug their geoserver configurations.